sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -307,9 +307,176 @@ case class StringLength(child: Expression) extends UnaryExpression with ExpectsI
     if (string == null) null else string.asInstanceOf[UTF8String].length
   }
 
-  override def toString: String = s"length($child)"
+  override def toString: String = s"LENGTH($child)"
 
   override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
     defineCodeGen(ctx, ev, c => s"($c).length()")
   }
 }
+
+/**
+ * Returns the numeric value of the first character of str.
+ */
+case class Ascii(child: Expression) extends UnaryExpression {
+  override def dataType: DataType = IntegerType
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (child.dataType != StringType) {
+      TypeCheckResult.TypeCheckFailure(
+        s"type of child expression in ASCII should be string, not ${child.dataType}")
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override def eval(input: InternalRow): Any = {
+    val string = child.eval(input)
+    if (string == null) {
+      null
+    } else {
+      val bytes = string.asInstanceOf[UTF8String].getBytes
+      if (bytes.length > 0) {
+        bytes(0).asInstanceOf[Int]
+      } else {
+        0
+      }
+    }
+  }
+
+  override def toString: String = s"ASCII($child)"
+}
+
+/**
+ * Converts the argument from binary to a base 64 string.
+ */
+case class Base64(child: Expression) extends UnaryExpression {
+  override def dataType: DataType = StringType
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (child.dataType != BinaryType) {
+      TypeCheckResult.TypeCheckFailure(
+        s"type of child expression in BASE64 should be binary, not ${child.dataType}")
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override def eval(input: InternalRow): Any = {
+    val bytes = child.eval(input)
+    if (bytes == null) {
+      null
+    } else {
+      UTF8String.fromBytes(
+        org.apache.commons.codec.binary.Base64.encodeBase64(
+          bytes.asInstanceOf[Array[Byte]]))
+    }
+  }
+
+  override def toString: String = s"BASE64($child)"
+}
+
+/**
+ * Converts the argument from a base 64 string to BINARY.
+ */
+case class UnBase64(child: Expression) extends UnaryExpression {
+  override def dataType: DataType = BinaryType
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (child.dataType != StringType) {
+      TypeCheckResult.TypeCheckFailure(
+        s"type of child expression in UNBASE64 should be string, not ${child.dataType}")
```

Comment:
Proabably `ExpectsInputTypes` and `AutoCastInputTypes` are slightly different, the former more like throws exception when the input types is not required, but later will convert the data types.
Let me know if you want to refactor that now, so we'd better suspend the job of rewriting the Hive UDF temporally.

One thing for sure, we can simplify the type checking, currently we almost have to put a description for the failure reason for most of the expressions.